### DISCUSSION TITLE
How do I run extra custom initialization after direnv? (ZSH)

### DISCUSSION THREAD

**Discussion by Hubro** [direnv/direnv#1318] [Q&A]
I'm using direnv with ZSH and I want to add some extra shell-specific initialization to my projects.

My first attempt was to add this to my `.zshrc`:

```zsh
# Optional extra init for projects using direnv
if [[ -n "$HUBRO_EXTRA_INIT" ]]; then
  eval "$HUBRO_EXTRA_INIT"
  unset HUBRO_EXTRA_INIT
fi
```

But since direnv only runs *after* the shell has initialized, this will never run.

**I need to be able to run my own ZSH function any time `direnv` has finished loading an environment.**

How can I do that in ZSH? Are there any official hooks I can use? (I assume no, since I've Googled with no success and had 3 different hallucinations suggested by ChatGPT.)

If not, are there any workarounds / hacks you can suggest?

---

My main motivation is to be able to source completions for project-specific CLIs, for example `eval "$(dev completions zsh)"` and set up simple aliases like `alias dev="python3 -m my_project.cli"`

**Comment by bdesham** [comment]
Here’s one idea. In your zshrc, somewhere after you set up direnv, put this:

```zsh
function process_extra_configuration() {
    if [[ -n "$HUBRO_EXTRA_INIT" ]]; then
        eval "$HUBRO_EXTRA_INIT"
    fi
}

chpwd_functions=($chpwd_functions process_extra_configuration)
```

This uses the same mechanism (`$chpwd_functions`) that direnv does in order to be called every time you change directories. By listing `process_extra_configuration` *after* all of the existing entries of the array—which includes direnv’s invocation function—you guarantee that your function will be run after direnv has had a chance to change your environment variables.

The only downside is that there’s not an easy way to undo this configuration once you leave the directory. You could maybe introduce a separate variable `$HUBRO_EXTRA_TEARDOWN` that undoes whatever `$HUBRO_EXTRA_INIT` did, but you would need to keep track of whether or not you were still in the original project directory, and only run the teardown code when you actually leave the directory.
